Traversable support for 'replace', 'merge' and 'sort'

SpacePossum opened the pull request:
Add traversable support to:
* replace
* merge
* sort

Other changes:
* Removes some not need checks
* ~~Make unit test skip if it cannot write to cache because of the user running the test~~
* Tests added for traversable support
* Doc updates

Deprecating:
I think the undocumented (and multibyte not supported) use of `replace(string, string)` can be removed in Twig 2.0.
Adding MB support and documentation of the feature was turned down (https://github.com/twigphp/Twig/pull/1618)
However removing it is a BC break (https://github.com/twigphp/Twig/pull/1445#issuecomment-48746926)

fabpot commented on doc/filters/merge.rst:
```diff
@@ -1,7 +1,7 @@
 =========
 
-The ``merge`` filter merges an array with another array:
+The ``merge`` filter merges an array or traversable with another array or traversable:
```
I would not mention it here as we are talking about Twig, not PHP. Instead, I would add a note at the bottom explaining the PHP implementation.

fabpot commented on doc/filters/sort.rst:
```diff
@@ -1,7 +1,7 @@
 ========
 
-The ``sort`` filter sorts an array:
+The ``sort`` filter sorts an array or traversable:
```
same here

fabpot commented on lib/Twig/Extension/Core.php:
```diff
@@ -550,6 +550,35 @@ function twig_date_converter(Twig_Environment $env, $date = null, $timezone = nu
+        $from = iterator_to_array($from);
+    } elseif (is_string($from) && is_string($to)) {
+        @trigger_error('Using "replace" with character by character replacement is deprecated and will be removed in Twig 2.x', E_USER_DEPRECATED);
+        return strtr($str, $from, $to);
```
missing blank line above, and blank line after should be removed

stof commented on lib/Twig/Extension/Core.php:
```diff
@@ -550,6 +550,36 @@ function twig_date_converter(Twig_Environment $env, $date = null, $timezone = nu
+
+        return strtr($str, $from, $to);
+
+    } elseif (false === is_array($from)) {
```
use ``!is_array`` here

stof commented on test/Twig/Tests/EnvironmentTest.php:
```diff
@@ -145,10 +145,17 @@ public function testExtensionsAreNotInitializedWhenRenderingACompiledTemplate()
+        $created = false;
+        if (!is_dir($dir)) {
+            if (false === $created = @mkdir($dir, 0777, true)) {
+                $this->markTestIncomplete(sprintf('Failed to create test cache directory at "%s".', $dir));
```
it should be marked as skipped, not incomplete. the test implementation is complete, but it cannot run on your system

stof commented on test/Twig/Tests/EnvironmentTest.php:
```diff
@@ -145,10 +145,17 @@ public function testExtensionsAreNotInitializedWhenRenderingACompiledTemplate()
+            }
+        }
+
+        if (false === @file_put_contents($cache, $twig->compileSource('{{ foo }}', 'index'))) {
```
please don't put the ``$twig->compileSource('{{ foo }}', 'index')`` expression inside the silenced statement, as we would loose any error triggered in the Twig code here

stof commented on test/Twig/Tests/Extension/CoreTest.php:
```diff
@@ -148,6 +148,92 @@ public function testTwigLast()
+    /**
+     * @dataProvider getTwigSArrayMergeData
+     */
+    public function testTwigSArrayMerge($arr1, $arr2, $expected)
```
What is SArray ?

SpacePossum commented:
@fabpot Thanks for the feedback, I've updated the docs. i hope you like those better.
@stof Good pointers, I've updated the code (`SArray` was a typo).

fabpot commented on doc/filters/merge.rst:
```diff
@@ -39,9 +39,10 @@ overridden.
 
-    Internally, Twig uses the PHP `array_merge`_ function.
+    Internally, Twig uses the PHP `array_merge`_ function. It supports
+    Traversable objects by transforming those to array.
```
To arrays

fabpot commented on lib/Twig/Extension/Core.php:
```diff
@@ -550,6 +550,36 @@ function twig_date_converter(Twig_Environment $env, $date = null, $timezone = nu
+                'The "replace" filter expects an array or "Traversable" as replace values, got "%s".',
+                is_object($from) ? get_class($from) : gettype($from)
+            )
+        );
```
should be on one line

fabpot commented on lib/Twig/Extension/Core.php:
```diff
@@ -550,6 +550,36 @@ function twig_date_converter(Twig_Environment $env, $date = null, $timezone = nu
+        @trigger_error('Using "replace" with character by character replacement is deprecated and will be removed in Twig 2.x', E_USER_DEPRECATED);
+
+        return strtr($str, $from, $to);
+
```
extra blank line should be removed

fabpot commented on lib/Twig/Extension/Core.php:
```diff
@@ -928,12 +966,18 @@ function twig_reverse_filter(Twig_Environment $env, $item, $preserveKeys = false
 {
+    if ($array instanceof Traversable) {
+        $array = iterator_to_array($array);
+    } elseif(!is_array($array)) {
```
missing space after elseif

fabpot commented on test/Twig/Tests/Extension/CoreTest.php:
```diff
@@ -148,6 +148,92 @@ public function testTwigLast()
+    /**
+     * @dataProvider getTwigReplaceFilterData
+     */
+    public function testTwigReplaceFilter($from, $replace, $expected)
```
Can you move those tests to the Fixtures directory like most of the other ones?

SpacePossum commented:
@fabpot I've moved the tests to Fixtures, please look again :)

fabpot commented on test/Twig/Tests/Fixtures/filters/replace.test:
```diff
@@ -1,8 +1,12 @@
+return array('traversable' => new ArrayObject(array('%this%' => 'foo', '%that%' => 'bar')))
 --EXPECT--
-I like foo and bar.
+I liké foo and bar.
```
I don't understand why you added an accent here.

SpacePossum replied:
So we test multi byte character support as well to prevent accidental regression in the future.

fabpot commented on test/Twig/Tests/EnvironmentTest.php:
```diff
@@ -145,10 +145,18 @@ public function testExtensionsAreNotInitializedWhenRenderingACompiledTemplate()
+        $created = false;
+        if (!is_dir($dir)) {
+            if (false === $created = @mkdir($dir, 0777, true)) {
+                $this->markTestSkipped(sprintf('Failed to create test cache directory at "%s".', $dir));
```
If the test cache dir cannot be created, I want a hard failure, not skipped tests.

SpacePossum replied:
ok, however I would like to note that is is a bit troublesome to switch users just to run the test.
The test assumes that who ever is running the test is allowed to create directories and folders.

stof replied:
well, being allowed to create files and folders in the system temp directory is a meaningful expectation IMO. If your user is not allowed to do that, I would consider that your have a broken system setup.

SpacePossum replied:
OK, please see https://github.com/SpacePossum/Twig/commit/309c9e7b4fe23370c28aef67ede475f9c6b96c7c :)

fabpot commented on test/Twig/Tests/EnvironmentTest.php:
```diff
@@ -165,6 +173,9 @@ public function testExtensionsAreNotInitializedWhenRenderingACompiledTemplate()
         unlink($cache);
+        if ($createdDir) {
+            rmdir($dir);
+        }
```
Can you please revert all changes in this file. They are not related to the goal of this PR and controversial :)

SpacePossum replied:
alright, done in 27fcd6e53f5e40f757fac2cce5f9e4d073cf7163 :)
I'll squash later because the history on the branch is becoming a bit of a mess ;)

fabpot commented on test/Twig/Tests/Fixtures/filters/sort.test:
```diff
@@ -3,8 +3,10 @@
 --EXPECT--
 14
 barfoo
+123
```
Missing blank line

fabpot replied:
Same for some other files as well

SpacePossum replied:
:+1: 
added

fabpot commented:
Thank you @SpacePossum.

SpacePossum commented:
thank you @fabpot & @stof
